### PR TITLE
net/dns: fix crash in tests

### PR DESCRIPTION
**Pull request by bradfitz**
Looks like #12346 as submitted with failing tests.

Updates #12346

### CODE REVIEW

**Review by irbekrm — APPROVED** [review]
Thank for the fix! I don't know this part of the code well- is it worth checking that it's not nil in the `disableSplitDNSOptimization()` function?

**Comment by bradfitz** [comment]
> is it worth checking that it's not nil  ...

Really these things should always be set now (health tracker, knobs, etc)

But seems like we're missing a panic-if-nil in the constructor. And docs on that field. I can try that first and see if anything breaks.

**Comment by bradfitz** [comment]
Oh, nevermind... controlknobs in that call path was optional and had comments about that.

The old knobs passsed in ended up setting:

```go
// forwarder forwards DNS packets to a number of upstream nameservers.
type forwarder struct {
...
	controlKnobs *controlknobs.Knobs // or nil
```

But then that same value was used for something else and assumed to be non-nil.

File: net/dns/manager.go [hunk 1]
```diff
@@ -52,11 +52,13 @@ type Manager struct {
 
 	resolver *resolver.Resolver
 	os       OSConfigurator
-	knobs    *controlknobs.Knobs
-	goos     string // if empty, gets set to runtime.GOOS
+	knobs    *controlknobs.Knobs // or nil
```
Comment (irbekrm) [inline net/dns/manager.go:55]: Nice

**Review by irbekrm — APPROVED** [review]

**Pull request merged**